Doc/webhook remidiation (#7509)
gardener/gardener: docs/usage/shoot_status.md

That's because most control-plane API calls are made with a client-side timeout of `30s`, so if a webhook has `timeoutSeconds=30`
the overall request might still fail as there is overhead in communication with the API server and potential other webhooks.
Generally, it's [best pratice](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#timeouts) to specify low timeouts in WebhookConfigs.

> As an effort to correct this common problem, the webhook remediator has been created. This feature simply checks whether a customer's webhook matches a set of rules, described [here](https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/matchers/matcher.go#L66-L180). If at least one of the rule matches, it will change the constraints (`HibernationPossible` & `MaintenancePreconditionsSatisfied`) statues.

Reviewer: Don't link the `master` branch - this will outdate. Rather use a perma-link.

Author: Is a permalink really better? For EG, if a rule changes, then the permalink is unaffected since it's made for a specific commit hash. Isn't it better to just leave the path to the file so that we won't have to change to doc for all future updates?

Reviewer: Fair enough, but then drop the line suffix from the link.

> As an effort to correct this common problem, the webhook remediator has been created. This feature simply checks whether a customer's webhook matches a set of rules, described [here](https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/matchers/matcher.go#L66-L180). If at least one of the rule matches, it will change the constraints (`HibernationPossible` & `MaintenancePreconditionsSatisfied`) statues.

Reviewer: ```suggestion
As an effort to correct this common problem, the webhook remediator has been created. This feature simply checks whether a customer's webhook matches a set of rules, described [here](https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/matchers/matcher.go#L66-L180). If at least one of the rule matches, it will change the constraints (`HibernationPossible` and `MaintenancePreconditionsSatisfied`) statuses.
```

> In most cases, you can avoid this by simply excluding the kube-system namespace from your webhook.

Reviewer: ```suggestion
In most cases, you can avoid this by simply excluding the `kube-system` namespace from your webhook:

```

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
webhooks:
  - name: my-webhook.example.com
    namespaceSelector:
      matchExpressions:
      - key: gardener.cloud/purpose
        operator: NotIn
        values:
          - kube-system
    rules:
      - operations: ["*"]
        apiGroups: [""]
        apiVersions: ["v1"]
        resources: ["pods"]
        scope: "Namespaced"
```

However, some other resources (mostly without namespaces) might still trigger the remediator, namely:
- endpoints
- nodes
- podsecuritypolicies
- clusterroles
- clusterrolebindings
- customresourcedefinitions
- apiservices
- certificatesigningrequests
- priorityclasses

> In these cases, please make sure that your `rules` don't overlap with one of those resources (and their subresources).

Reviewer: I don't understand this sentence - what does it mean/what do I need to do when I have a webhook for `ClusterRole`s, for example?

Author: Just changed the sentence as it was indeed not clear enough. Basically, my point was : if one of the non-namespaced resources triggers the remediator, simply remove that resource from the webhook's rule.

Reviewer: OK, but what if I can't remove it because my software relies on this webhook to be present?

Author: Well AFAIK, that's how the webhook remediator works (I guess that's a limitation)... I don't think there is another way around it. Are you aware of another way to work around that for non-namespaced resources ?

Reviewer: In such (rather rare/special, I assume) cases, end-users should probably label their webhooks with `remediation.webhook.shoot.gardener.cloud/exclude=true` as explained in the document as well to not be bothered by it. However, we should make clear that this should only be used in those cases and not to "quickly silence/stop the auto-remediation because it bothers me".

Author: Good point! I just revamped the last section to make it clearer.

> You can also find help from the [kubernetes documentation](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#best-practices-and-warnings)

Reviewer: ```suggestion
You can also find help from the [Kubernetes documentation](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#best-practices-and-warnings)
```
